language/stdlib/modules/Roles.move

```diff
@@ -381,61 +380,238 @@ module Roles {
+    }
+    spec module {
+        apply ValidatorOperatorRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    spec schema ParentVASPRoleMatchesRoleId {
+        invariant module forall addr: address where exists<Privilege<ParentVASPRole>>(addr):
+             spec_has_role_id(addr)
+                && (spec_get_role_id(addr) == SPEC_PARENT_VASP_ROLE_ID());
+
+        invariant module forall addr: address where spec_has_role_id(addr):
+            (spec_get_role_id(addr) == SPEC_PARENT_VASP_ROLE_ID())
+                ==> exists<Privilege<ParentVASPRole>>(addr);
+    }
+    spec module {
+        apply ParentVASPRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    spec schema ChildVASPRoleMatchesRoleId {
+        invariant module forall addr: address where exists<Privilege<ChildVASPRole>>(addr):
+             spec_has_role_id(addr)
+                && (spec_get_role_id(addr) == SPEC_CHILD_VASP_ROLE_ID());
+
+        invariant module forall addr: address where spec_has_role_id(addr):
+            (spec_get_role_id(addr) == SPEC_CHILD_VASP_ROLE_ID())
+                ==> exists<Privilege<ChildVASPRole>>(addr);
+    }
+    spec module {
+        apply ChildVASPRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    spec schema UnhostedRoleMatchesRoleId {
+        invariant module forall addr: address where exists<Privilege<UnhostedRole>>(addr):
+             spec_has_role_id(addr)
+                && (spec_get_role_id(addr) == SPEC_UNHOSTED_ROLE_ID());
+
+        invariant module forall addr: address where spec_has_role_id(addr):
+            (spec_get_role_id(addr) == SPEC_UNHOSTED_ROLE_ID())
+                ==> exists<Privilege<UnhostedRole>>(addr);
+    }
+    spec module {
+        apply UnhostedRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    /// ## Role persistence
+
+    /// **Informally:** Once an account at address `A` is granted a role `R` it
+    /// will remain an account with role `R` for all time.
+    spec schema RoleIdPersists {
+        ensures forall addr: address where old(spec_has_role_id(addr)) :
+            spec_has_role_id(addr) && (old(spec_get_role_id(addr)) == spec_get_role_id(addr));
+    }
+
+    spec module {
+        apply RoleIdPersists to *<T>, *;
+    }
+
+    /// ## Privilege extraction
+
+/// >TODO: This has not been proved yet and may be wrong, or need further
+/// strengthening.  Or, it may just be too hard for the prover.
+/// The capability returned from extract_privilege_to_capability is free-floating,
+/// so we have to quantify over members of the type, not over addresses or something.
+
+    spec schema IsExtractedIffCapabilityExists {
+        /// **Informally:** No capabilities for a privilege witness at a given
+        /// address exist until the privilege exists.
+        /// > NOTE: This verifies by itself when applied to everything.
+        invariant forall priv: type, addr: address where !exists<Privilege<priv>>(addr):
```

Comment:
It's probably more clear to state the contrapositive here:
`forall priv: type, c: Capability<priv>: exists<Privilege<priv>>(c.owner_address)`.
This way we are directly saying that capabilities must come from accounts with the corresponding privileges.